kv/kvserver: TestRebalancingAndCrossRegionZoneSnapshotMetrics failed

kv/kvserver.TestRebalancingAndCrossRegionZoneSnapshotMetrics [failed](https://tanzanite.cluster.engflow.com/invocations/default/29be128b-bbb2-4b8c-89de-c0a0168fe550?testReportRun=3&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [d18eb683b2759fd8814dacf0baa913f596074a17](https://github.com/cockroachdb/cockroach/commits/d18eb683b2759fd8814dacf0baa913f596074a17):


```
        	            	expected: map[string]kvserver_test.snapshotBytesMetrics{"":kvserver_test.snapshotBytesMetrics{sentBytes:0, rcvdBytes:567}, ".cross-region":kvserver_test.snapshotBytesMetrics{sentBytes:0, rcvdBytes:567}, ".cross-zone":kvserver_test.snapshotBytesMetrics{sentBytes:0, rcvdBytes:0}, ".rebalancing":kvserver_test.snapshotBytesMetrics{sentBytes:0, rcvdBytes:567}, ".recovery":kvserver_test.snapshotBytesMetrics{sentBytes:0, rcvdBytes:0}}
        	            	actual  : map[string]kvserver_test.snapshotBytesMetrics{"":kvserver_test.snapshotBytesMetrics{sentBytes:0, rcvdBytes:559}, ".cross-region":kvserver_test.snapshotBytesMetrics{sentBytes:0, rcvdBytes:559}, ".cross-zone":kvserver_test.snapshotBytesMetrics{sentBytes:0, rcvdBytes:0}, ".rebalancing":kvserver_test.snapshotBytesMetrics{sentBytes:0, rcvdBytes:559}, ".recovery":kvserver_test.snapshotBytesMetrics{sentBytes:0, rcvdBytes:0}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -3,3 +3,3 @@
        	            	   sentBytes: (int64) 0,
        	            	-  rcvdBytes: (int64) 567
        	            	+  rcvdBytes: (int64) 559
        	            	  },
        	            	@@ -7,3 +7,3 @@
        	            	   sentBytes: (int64) 0,
        	            	-  rcvdBytes: (int64) 567
        	            	+  rcvdBytes: (int64) 559
        	            	  },
        	            	@@ -15,3 +15,3 @@
        	            	   sentBytes: (int64) 0,
        	            	-  rcvdBytes: (int64) 567
        	            	+  rcvdBytes: (int64) 559
        	            	  },
        	Test:       	TestRebalancingAndCrossRegionZoneSnapshotMetrics/first_receiver
    --- FAIL: TestRebalancingAndCrossRegionZoneSnapshotMetrics/first_receiver (0.00s)
=== RUN   TestRebalancingAndCrossRegionZoneSnapshotMetrics/sender
    replica_learner_test.go:2371: 
        	Error Trace:	pkg/kv/kvserver/replica_learner_test.go:2371
        	Error:      	Not equal: 
        	            	expected: map[string]kvserver_test.snapshotBytesMetrics{"":kvserver_test.snapshotBytesMetrics{sentBytes:1356, rcvdBytes:0}, ".cross-region":kvserver_test.snapshotBytesMetrics{sentBytes:567, rcvdBytes:0}, ".cross-zone":kvserver_test.snapshotBytesMetrics{sentBytes:789, rcvdBytes:0}, ".rebalancing":kvserver_test.snapshotBytesMetrics{sentBytes:1356, rcvdBytes:0}, ".recovery":kvserver_test.snapshotBytesMetrics{sentBytes:0, rcvdBytes:0}}
        	            	actual  : map[string]kvserver_test.snapshotBytesMetrics{"":kvserver_test.snapshotBytesMetrics{sentBytes:1348, rcvdBytes:0}, ".cross-region":kvserver_test.snapshotBytesMetrics{sentBytes:559, rcvdBytes:0}, ".cross-zone":kvserver_test.snapshotBytesMetrics{sentBytes:789, rcvdBytes:0}, ".rebalancing":kvserver_test.snapshotBytesMetrics{sentBytes:1348, rcvdBytes:0}, ".recovery":kvserver_test.snapshotBytesMetrics{sentBytes:0, rcvdBytes:0}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -2,3 +2,3 @@
        	            	  (string) "": (kvserver_test.snapshotBytesMetrics) {
        	            	-  sentBytes: (int64) 1356,
        	            	+  sentBytes: (int64) 1348,
        	            	   rcvdBytes: (int64) 0
        	            	@@ -6,3 +6,3 @@
        	            	  (string) (len=13) ".cross-region": (kvserver_test.snapshotBytesMetrics) {
        	            	-  sentBytes: (int64) 567,
        	            	+  sentBytes: (int64) 559,
        	            	   rcvdBytes: (int64) 0
        	            	@@ -14,3 +14,3 @@
        	            	  (string) (len=12) ".rebalancing": (kvserver_test.snapshotBytesMetrics) {
        	            	-  sentBytes: (int64) 1356,
        	            	+  sentBytes: (int64) 1348,
        	            	   rcvdBytes: (int64) 0
        	Test:       	TestRebalancingAndCrossRegionZoneSnapshotMetrics/sender
    --- FAIL: TestRebalancingAndCrossRegionZoneSnapshotMetrics/sender (0.00s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRebalancingAndCrossRegionZoneSnapshotMetrics.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45490

--- a/pkg/kv/kvserver/replica_learner_test.go
+++ b/pkg/kv/kvserver/replica_learner_test.go
@@ -2301,6 +2301,17 @@ func TestRebalancingAndCrossRegionZoneSnapshotMetrics(t *testing.T) {
 	defer tc.Stopper().Stop(ctx)
 
 	scratchStartKey := tc.ScratchRange(t)
+	desc := tc.LookupRangeOrFatal(t, scratchStartKey)
+	// Wait for the expiration lease to upgrade to an epoch or leader lease.
+	// Otherwise, the lease upgrade may race with the snapshot calculation below
+	// and result in a different size snapshot than expected. For epoch leases
+	// this is actually not necessary because the first lease that's proposed is
+	// an epoch lease. For leader leases, however, the range starts off with no
+	// leader so the first lease that's proposed is an expiration lease, which
+	// gets upgraded to a leader lease once a leader is elected.
+	if !kvserver.ExpirationLeasesOnly.Get(&tc.Server(0).ClusterSettings().SV) {
+		tc.WaitForLeaseUpgrade(ctx, t, desc)
+	}
 	// sendSnapshotFromServer is a testing helper that sends a learner snapshot
 	// from server[0] to server[serverIndex] and returns the expected size (in
 	// bytes) of the snapshot sent.

